# I. Data Coding and Merging

In [1]:
# Connect to Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [61]:
# Load packages
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [62]:
# Load game logs data
Teams = pd.read_excel("/content/drive/MyDrive/Sports Performance Analytics Michigan/Course 2 - Moneyball and Beyond/Week 1 Content/Week 1 Assignment/Game logs 1999-2003 (1).xlsx")

In [63]:
# Create a dummy variable for home wins (1 or 0) and visitor wins, and create the ‘year’ variable.
Teams['hwin'] = np.where(Teams['home_score'] > Teams['visitor_score'],1,0)
Teams['awin'] = np.where(Teams['home_score'] < Teams['visitor_score'],1,0)
Teams['year'] = Teams['date'].astype(str).str[0:4]

In [64]:
# Create a dataframe that sums the statistics you need by season (year) and by home team. Rename the column for the home team as ‘Team’.
Teamshome = Teams.groupby(['home','year'])[['visitor_ab', 'visitor_h', 'visitor_2b', 'visitor_3b', 'visitor_hr', 'visitor_sf',
                                           'visitor_hbp','visitor_bb','home_ab', 'home_h', 'home_2b', 'home_3b',
                                           'home_hr', 'home_sf', 'home_hbp', 'home_bb', 'hwin','awin']].sum()
Teamshome = Teamshome.reset_index()
Teamshome = Teamshome.rename(columns = {"home": "Team"})
Teamshome

,Team,year,visitor_ab,visitor_h,visitor_2b,visitor_3b,visitor_hr,visitor_sf,visitor_hbp,visitor_bb,home_ab,home_h,home_2b,home_3b,home_hr,home_sf,home_hbp,home_bb,hwin,awin
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39


In [65]:
# Create a dataframe that sums the statistics you need by season (year) and by visiting team. Rename the column for the visiting team as ‘Team’.
Teamsaway = Teams.groupby(['year','visitor'])[['visitor_ab', 'visitor_h', 'visitor_2b', 'visitor_3b',
                                           'visitor_hr', 'visitor_sf', 'visitor_hbp', 'visitor_bb',
                                           'home_ab', 'home_h', 'home_2b', 'home_3b', 'home_hr',
                                           'home_sf', 'home_hbp', 'home_bb','hwin','awin']].sum()

Teamsaway = Teamsaway.reset_index()
Teamsaway = Teamsaway.rename(columns = {"visitor": "Team"})
Teamsaway

,year,Team,visitor_ab,visitor_h,visitor_2b,visitor_3b,visitor_hr,visitor_sf,visitor_hbp,visitor_bb,home_ab,home_h,home_2b,home_3b,home_hr,home_sf,home_hbp,home_bb,hwin,awin
0,1999,ANA,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33
1,1999,ARI,2928,812,147,15,115,29,31,303,2751,694,123,12,92,14,22,287,33,48
2,1999,ATL,2901,777,151,14,111,19,31,322,2772,724,131,9,67,22,16,286,34,47
3,1999,BAL,2905,810,163,12,105,23,25,300,2638,708,157,19,97,22,19,318,44,37
4,1999,BOS,2857,758,157,22,96,27,25,292,2704,674,134,10,88,23,23,243,36,45
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,2003,SFN,2752,689,147,10,98,18,26,287,2590,657,124,12,75,26,23,285,37,43
146,2003,SLN,2914,781,159,21,111,30,34,300,2831,815,180,16,117,27,37,276,44,37
147,2003,TBA,2859,749,140,17,81,29,27,188,2663,719,163,17,100,30,32,317,54,27
148,2003,TEX,2877,708,133,13,99,15,41,242,2719,770,154,17,103,34,24,319,53,28


In [66]:
# Merge the two dataframes on ‘year’ and ‘Team’ and create a variable for the total number of wins in the season for each team
Teams2 = pd.merge(Teamshome, Teamsaway, on =['Team','year'])
pd.set_option('display.max_columns', 50)
display(Teams2)

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36


In [67]:
Teams2['wins'] = Teams2['hwin_x']+Teams2['awin_y']
display(Teams2)

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y,wins
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33,70
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36,82
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36,75
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45,99
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32,77
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44,84
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38,83
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40,80
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36,78


# II. Calculating Performance Metrics

In [70]:
# Create batting average for and batting average against as the sum the relevant statistics for each team as home team and visiting team.
Teams2['BAFOR'] = (Teams2['home_h_x']+Teams2['visitor_h_y'])/(Teams2['home_ab_x']+Teams2['visitor_ab_y'])
Teams2['BAAGN'] = (Teams2['home_h_y']+Teams2['visitor_h_x'])/(Teams2['home_ab_y']+Teams2['visitor_ab_x'])

In [71]:
# Create the win percentage variable for each team by generating a count of games played and dividing total wins by total games played.
TeamsGh = Teams.groupby(['year','home'])['hwin'].count()
TeamsGh = TeamsGh.reset_index()
TeamsGh = TeamsGh.rename(columns = {"home": 'Team'})

TeamsGa = Teams.groupby(['year','visitor'])['awin'].count()
TeamsGa = TeamsGa.reset_index()
TeamsGa = TeamsGa.rename(columns = {"visitor": 'Team'})

TeamsG = pd.merge(TeamsGh, TeamsGa, on =['Team','year'])
TeamsG['Games'] = TeamsG['hwin']+TeamsG['awin']
TeamsG = TeamsG[['year','Team','Games']]
TeamsG

,year,Team,Games
0,1999,ANA,162
1,1999,ARI,162
2,1999,ATL,162
3,1999,BAL,162
4,1999,BOS,162
...,...,...,...
145,2003,SFN,161
146,2003,SLN,162
147,2003,TBA,162
148,2003,TEX,162


In [72]:
Teams2 = pd.merge(Teams2, TeamsG, on =['Team','year'])
Teams2

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y,wins,BAFOR,BAAGN,Games
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33,70,0.255552,0.268760,162
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36,82,0.279673,0.273489,162
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36,75,0.260674,0.262615,162
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45,99,0.282318,0.246653,162
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32,77,0.268453,0.261027,162
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44,84,0.280043,0.280298,162
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38,83,0.275145,0.284883,162
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40,80,0.262935,0.274916,162
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36,78,0.261064,0.269389,162


In [73]:
Teams2['wpc'] = Teams2['wins']/Teams2['Games']
Teams2

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y,wins,BAFOR,BAAGN,Games,wpc
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33,70,0.255552,0.268760,162,0.432099
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36,82,0.279673,0.273489,162,0.506173
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36,75,0.260674,0.262615,162,0.462963
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45,99,0.282318,0.246653,162,0.611111
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32,77,0.268453,0.261027,162,0.475309
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44,84,0.280043,0.280298,162,0.518519
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38,83,0.275145,0.284883,162,0.512346
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40,80,0.262935,0.274916,162,0.493827
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36,78,0.261064,0.269389,162,0.481481


In [74]:
# Using a similar process to part 2), calculate each team’s home and away win percentage separately for each season.
TeamsGth = Teams.groupby(['year','home'])['hwin'].count()
TeamsGth = TeamsGth.reset_index()
TeamsGth = TeamsGth.rename(columns = {"home": 'Team',"hwin":'home_games'})

TeamsGta = Teams.groupby(['year','visitor'])['awin'].count()
TeamsGta = TeamsGta.reset_index()
TeamsGta = TeamsGta.rename(columns = {"visitor": 'Team',"awin":'away_games'})

TeamsGt = pd.merge(TeamsGth, TeamsGta, on =['Team','year'])
TeamsGt

,year,Team,home_games,away_games
0,1999,ANA,81,81
1,1999,ARI,81,81
2,1999,ATL,81,81
3,1999,BAL,81,81
4,1999,BOS,81,81
...,...,...,...,...
145,2003,SFN,81,80
146,2003,SLN,81,81
147,2003,TBA,81,81
148,2003,TEX,81,81


In [75]:
# Merge each team’s home and away win percentage for each season into one data frame.
Teams2 = pd.merge(Teams2, TeamsGt, on =['Team','year'])
Teams2

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y,wins,BAFOR,BAAGN,Games,wpc,home_games,away_games
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33,70,0.255552,0.268760,162,0.432099,81,81
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36,82,0.279673,0.273489,162,0.506173,81,81
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36,75,0.260674,0.262615,162,0.462963,81,81
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45,99,0.282318,0.246653,162,0.611111,81,81
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32,77,0.268453,0.261027,162,0.475309,82,80
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44,84,0.280043,0.280298,162,0.518519,81,81
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38,83,0.275145,0.284883,162,0.512346,81,81
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40,80,0.262935,0.274916,162,0.493827,82,80
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36,78,0.261064,0.269389,162,0.481481,81,81


In [76]:
Teams2['hwpc'] = Teams2['hwin_x']/Teams2['home_games']
Teams2['awpc'] = Teams2['awin_y']/Teams2['away_games']
Teams2

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y,wins,BAFOR,BAAGN,Games,wpc,home_games,away_games,hwpc,awpc
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33,70,0.255552,0.268760,162,0.432099,81,81,0.456790,0.407407
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36,82,0.279673,0.273489,162,0.506173,81,81,0.567901,0.444444
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36,75,0.260674,0.262615,162,0.462963,81,81,0.481481,0.444444
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45,99,0.282318,0.246653,162,0.611111,81,81,0.666667,0.555556
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32,77,0.268453,0.261027,162,0.475309,82,80,0.548780,0.400000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44,84,0.280043,0.280298,162,0.518519,81,81,0.493827,0.543210
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38,83,0.275145,0.284883,162,0.512346,81,81,0.555556,0.469136
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40,80,0.262935,0.274916,162,0.493827,82,80,0.487805,0.500000
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36,78,0.261064,0.269389,162,0.481481,81,81,0.518519,0.444444


In [77]:
# Calculate WPCT diff defined as home win percentage – away win percentage for every team in each season.
Teams2['WPCT_diff'] = Teams2['hwpc']-Teams2['awpc']
Teams2

,Team,year,visitor_ab_x,visitor_h_x,visitor_2b_x,visitor_3b_x,visitor_hr_x,visitor_sf_x,visitor_hbp_x,visitor_bb_x,home_ab_x,home_h_x,home_2b_x,home_3b_x,home_hr_x,home_sf_x,home_hbp_x,home_bb_x,hwin_x,awin_x,visitor_ab_y,visitor_h_y,visitor_2b_y,visitor_3b_y,visitor_hr_y,visitor_sf_y,visitor_hbp_y,visitor_bb_y,home_ab_y,home_h_y,home_2b_y,home_3b_y,home_hr_y,home_sf_y,home_hbp_y,home_bb_y,hwin_y,awin_y,wins,BAFOR,BAAGN,Games,wpc,home_games,away_games,hwpc,awpc,WPCT_diff
0,ANA,1999,2832,762,151,10,105,39,27,322,2666,669,118,13,74,21,23,267,37,44,2828,735,130,9,84,21,20,244,2645,710,134,15,72,26,29,302,48,33,70,0.255552,0.268760,162,0.432099,81,81,0.456790,0.407407,0.049383
1,ANA,2000,2867,800,144,8,113,24,15,320,2742,788,146,11,130,23,24,315,46,35,2886,786,163,23,106,20,23,293,2742,734,132,15,115,29,21,342,45,36,82,0.279673,0.273489,162,0.506173,81,81,0.567901,0.444444,0.123457
2,ANA,2001,2873,772,165,6,86,20,39,259,2736,731,122,12,86,29,30,251,39,42,2815,716,153,14,72,24,47,243,2656,680,120,13,82,19,25,266,45,36,75,0.260674,0.262615,162,0.462963,81,81,0.481481,0.444444,0.037037
3,ANA,2002,2799,685,132,5,70,29,22,260,2757,769,155,21,71,27,34,228,54,27,2921,834,178,11,81,37,40,234,2654,660,116,14,99,30,27,249,36,45,99,0.282318,0.246653,162,0.611111,81,81,0.666667,0.555556,0.111111
4,ANA,2003,2868,743,142,9,84,17,39,245,2685,717,135,17,68,22,27,226,45,37,2802,756,141,16,82,28,29,250,2664,701,118,18,106,25,37,241,48,32,77,0.268453,0.261027,162,0.475309,82,80,0.548780,0.400000,0.148780
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145,TOR,1999,2891,817,190,11,93,31,25,289,2731,765,173,6,96,21,44,283,40,41,2911,815,164,8,116,24,32,295,2753,765,135,22,98,26,28,286,37,44,84,0.280043,0.280298,162,0.518519,81,81,0.493827,0.543210,-0.049383
146,TOR,2000,2910,819,182,11,92,19,30,279,2766,768,178,9,134,12,33,255,45,36,2911,794,150,12,110,22,27,271,2759,796,156,12,103,21,34,281,43,38,83,0.275145,0.284883,162,0.512346,81,81,0.555556,0.469136,0.086420
147,TOR,2001,2911,800,207,17,84,25,46,264,2809,757,146,14,94,31,37,256,40,42,2854,732,141,22,101,12,37,214,2738,753,150,9,81,18,30,226,40,40,80,0.262935,0.274916,162,0.493827,82,80,0.487805,0.500000,-0.012195
148,TOR,2002,2838,749,171,10,89,28,37,288,2729,734,164,20,102,24,30,258,42,39,2852,723,141,18,85,33,23,264,2745,755,141,19,88,24,34,302,45,36,78,0.261064,0.269389,162,0.481481,81,81,0.518519,0.444444,0.074074


In [79]:
# Create a variable AVG Diff = Batting Average For – Batting Average Against
Teams2['AVG_Diff'] = Teams2['BAFOR']-Teams2['BAAGN']

# III. Running Regressions

In [83]:
# Run an OLS (Ordinary Least Squares) regression of win percentage on batting average for and batting average against
import statsmodels.formula.api as smf
WinBA_lm = smf.ols(formula = 'wpc ~ BAFOR + BAAGN', data=Teams2).fit()
WinBA_lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    wpc   R-squared:                       0.720
Model:                            OLS   Adj. R-squared:                  0.716
Method:                 Least Squares   F-statistic:                     188.6
Date:                Sat, 08 Aug 2026   Prob (F-statistic):           2.60e-41
Time:                        20:48:42   Log-Likelihood:                 265.67
No. Observations:                 150   AIC:                            -525.3
Df Residuals:                     147   BIC:                            -516.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.6016      0.099      6.069      0.000       0.406       0.797
BAFOR          3.7863      0.300     12.619      0.000       3.193       4.379
BAAGN         -4.1691      0.258    -16.141      0.000      -4.680      -3.659
==============================================================================
Omnibus:                        0.210   Durbin-Watson:                   1.530
Prob(Omnibus):                  0.900   Jarque-Bera (JB):                0.033
Skew:                           0.005   Prob(JB):                        0.983
Kurtosis:                       3.073   Cond. No.                         97.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We see a R-squared of 0.720 and coefficients from batting average for and against being 3.7863 and -4.1691 respectively. This means that a one point increase in batting average for will increase win percentage by 3.7863 points and a one point increase in batting average against will decrease the win percentage by 4.1691 points.

In [84]:
# Run an OLS (Ordinary Least Squares) regression of win percentage on batting average for and batting average against,constraining the coefficient on each variable to be equal and opposite.
WinBAcon_lm = smf.ols(formula = 'wpc ~ I(BAFOR - BAAGN)', data=Teams2).fit()
WinBAcon_lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    wpc   R-squared:                       0.718
Model:                            OLS   Adj. R-squared:                  0.716
Method:                 Least Squares   F-statistic:                     376.0
Date:                Sat, 08 Aug 2026   Prob (F-statistic):           1.81e-42
Time:                        20:51:12   Log-Likelihood:                 265.13
No. Observations:                 150   AIC:                            -526.3
Df Residuals:                     148   BIC:                            -520.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.4997      0.003    147.125      0.000       0.493       0.506
I(BAFOR - BAAGN)     4.0099      0.207     19.390      0.000       3.601       4.419
==============================================================================
Omnibus:                        0.360   Durbin-Watson:                   1.507
Prob(Omnibus):                  0.835   Jarque-Bera (JB):                0.108
Skew:                          -0.003   Prob(JB):                        0.948
Kurtosis:                       3.131   Cond. No.                         60.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We see a R-squared of 0.718 and coefficients from batting average for and against being 4.0099 and -4.0099 respectively. This means that a one point increase in batting average for will increase win percentage by 4.0099 points and a one point increase in batting average against will decrease the win percentage by 4.0099 points.